### ISSUE TITLE
Editor window startup is slow

### ISSUE THREAD

**Issue by renatodex** [atom/atom#2654]
Hi, i have used Textmate for a loong long time. But also as a big fan of new things, i decided to try Atom for curiosity purposes.

It turns out that i found Atom a great platform for day-by-day development, and im using it Atom now. But theres one thing annoying me:

I usually open Atom using the Shell command "atom [filename]", and compared to "mate [filename]" (the same command Textmate use for open files using Shell), Atom is incredible slow (i feel i have to wait twice of the time, like 1s~2s).
I know this performance also is related to machine specs, processors, memory and etc. But its the exactly same task executed for different tools. 

I can record a video showing that behavior if you like.
Also, if this is a un-prioritize known issue, i can submit a pull request if theres noone working on it. I just need to know where to start looking.

][`s

**Comment by helmus** [comment]
Just thought I would drop this here, cold boots are always long on average but this one took extra long. This is with the default packages and I'm using version 0.152.0

![image](https://cloud.githubusercontent.com/assets/1042463/5626716/72685a1a-958f-11e4-9a13-15ba1219d3d1.png)

**Comment by lee-dohm** [comment]
@helmus You probably want to update to the latest version. I assume that you're on Windows since I've seen a lot of Windows users with old versions due to the Installer change:

http://blog.atom.io/2014/12/10/a-windows-installer-and-updater.html

**Comment by helmus** [comment]
@lee-dohm Turns out I actually am running the latest version. I got confused because it says 0.152.0 on the release notes as latest version. ( not sure if that's a bug ? ). When I click "Restart And Update" it says: "you are running the latest version 0.166.0".
Anyway, cold boots are still very very slow. It seems to improve somewhat after restarting once or twice, pretty strange really.

![image](https://cloud.githubusercontent.com/assets/1042463/5630017/34873a10-95ba-11e4-9b0e-1be5e80042d8.png)

**Comment by Mulletfingers999** [comment]
:+1: my only complaint

**Comment by Sigmanor** [comment]
First launch:
![](http://i.imgur.com/68lL5BP.png)

Second and others launches.
![](http://i.imgur.com/i1VWXR0.png)

Windows 8.1
2.60 GHz i5-3230M 
6 GB RAM
1TB HDD

**Comment by benogle** [comment]
That first launch is insane. It took over half a second just to load/render the theme css for each theme. Was your machine loaded up during that time.

The second run is more reasonable. It would be considerably faster without emmet or the color picker.

**Comment by 50Wliu** [comment]
@benogle, here's one from a fully warmed up machine (it had been on for ~3 hours when this was taken):
![sad](https://cloud.githubusercontent.com/assets/2766036/5908759/57d20acc-a578-11e4-9796-868152fb12fe.png)

**Comment by kevinsawicki** [comment]
@50Wliu solid state?

**Comment by 50Wliu** [comment]
Nope, normal hard drive.

**Comment by zcbenz** [comment]
One possible way to fix the speed problem on HHD is to use [asar](https://github.com/atom/atom-shell/blob/master/docs/tutorial/application-packaging.md#generating-asar-archive) to archive all the source code files into one file, so when loading Atom will only read one file instead of reading nearly 600 files.

**Comment by dustincaruso** [comment]
- just wanted to add my 2¢...

- "New Window" (blank)
- no packages or plugins (fairly new install)
- on an iMac w/ Quad Core 2.8 i7 w/ 16GB RAM

= 3.5 seconds
= reaaalllly slow

**Comment by bpasero** [comment]
@kevinsawicki you say:

> Pretty much, we've never really profiled Atom shell load time deeply and seen about possible 
> optimizations there, but that would probably be a max speedup of 100ms since the Shell load time is 
> usually never the bad one, the Window load time is and Atom is in complete control over that.

I did some testing with Atom shell to understand how long it takes until my main script gets control. For that I am printing the unix timestamp before launching Atom shell and then within I just have a single console.log printing the unix timestamp at that time. Here are my times on a 2014 MacBookPro running windows (Boot camp):

* Cold: 515ms
* Warm: 140ms

Compared to doing the same with a plain node.js script:

* Cold: 140ms
* Warm: 50ms

I am wondering what takes so long in Atom shell. I am not creating any renderer processes, it is just a main.js with a single console.log statement. Any ideas?

**Comment by kevinsawicki** [comment]
> I am wondering what takes so long in Atom shell.

Atom Shell performance discussions should probably be taken to a new issue over on that repo.

> Compared to doing the same with a plain node.js script:

Atom Shell has more going on than just node, it also loads Chromium, so comparing Atom Shell to other native `.app` launch times is probably better to do.

**Comment by procedural** [comment]
![](http://i.imgur.com/53AVPra.png)

Xubuntu 14.04, `atom --safe`, non-solid-state hard drive.

Looking at @Macsch15's screenshots I assume SSD really make big difference?

**Comment by procedural** [comment]
Ok, same setup but instead of open source Mesa 10.5.0 video driver I've installed Catalyst Omega:

![](http://i.imgur.com/0W7Ih84.png)

I'm using a notebook with fairly weak integrated videocard.

By the way in both cases "Use Hardware Acceleration" flag doesn't affect these timings.

**Comment by shahril96** [comment]
For me, it's fast enough. By the way, sometimes I also got 1000+ ms but it don't bother me much.

![atomstartup](https://cloud.githubusercontent.com/assets/2675341/6205632/0e49f840-b5b0-11e4-9b42-0ca3f34c67da.png)

**Comment by 50Wliu** [comment]
I wish mine looked like yours :cry:.  ;P

**Comment by cthielen** [comment]
I'm not sure if this belongs in the same issue but using the 'atom' command while Atom is already open also temporarily shows a second Atom icon in the dock that optimally wouldn't be there.

**Comment by muchweb** [comment]
>I'm not sure if this belongs in the same issue

@cthielen, I think you are better creating new issue for this.

**Comment by kylejson** [comment]
![screen shot 2015-03-19 at 12 24 57 pm](https://cloud.githubusercontent.com/assets/1666767/6735193/0b687764-ce33-11e4-8cb4-ce6e06128b96.png)

**Comment by edevil** [comment]
Big improvements!

![screen shot 2015-04-15 at 13 19 08](https://cloud.githubusercontent.com/assets/12426/7158441/09e8a3ce-e372-11e4-8f54-2d6f303583d4.png)

**Comment by ceyhunkerti** [comment]
We are about to celebrate first year of this issue and it is still the one that is much stops users from adopting atom.

**Comment by Hollowbean** [comment]
+1, side note... Microsofts Visual Studio Code solves this for atom. Maybe someone can look at what they did and implement it back into the main atom branch?

**Comment by edevil** [comment]
+1 Good call. Hadn't noticed it.

**Comment by iclanzan** [comment]
@Hollowbean yeah, I don’t know how Microsoft did it, but VS Code is really fast to open up and has no problem handling large (and even minified) files. Tried editing a 10MB file without a problem.

**Comment by mark-hahn** [comment]
>  I don’t know how Microsoft did it

I assume they did not design for hackability.  They said they would offer
"plug-ins" in the future.  Atom was designed to
have accessible components.  And we don't know how much binary microsoft
added to the system.

Atom has some serious plans to speed it up.  If they pull that off it will
be the best of all worlds.  Fast and hackable.

**Comment by benjiro** [comment]
Finding huge improvements after the last few builds. Keep up the good work!

![image](https://cloud.githubusercontent.com/assets/2031163/7553092/5889f0e0-f723-11e4-8b13-99ae7b68039a.png)

**Comment by psychok7** [comment]
version 1.0 on Ubuntu 14.04 x64 it took about 4/5 seconds to load up 1 small file (no extensions) on a i7 8GB machine.

Loading performance definitely needs some love.

**Comment by xpol** [comment]
Yes, more love please. It's and editor not and IDE.

**Comment by Gabriel-p** [comment]
I'm definitely seeing a performance improvement with v1.0:

![captura de pantalla de 2015-06-27 01 12 39](https://cloud.githubusercontent.com/assets/6054951/8390416/ae83f8ae-1c69-11e5-90e1-f272ee395cf1.png)

not nearly as fast as `Sublime` yet, but slowly getting there.

**Comment by Sigmanor** [comment]
![](http://i.imgur.com/ednXjpQ.png)

**Comment by yeluolei** [comment]
start up too slow

**Comment by ben-sussman-zocdoc** [comment]
:+1: This is the issue that is preventing me from adopting Atom as my primary editor. I experience about a ~4s latency before the text of a tiny `misc.txt` file is rendered. New tabs and new files are fast, but from completely closed & cold to displaying text is about 4s. I'm am using a completely vanilla `atom.io` install on an 2.7Ghz i7 rMBP with 16gig of Ram.

**Comment by mnquintana** [comment]
> completely closed & cold to displaying text is about 4s

Is that the exact startup time? You can find out the exact startup time by opening Timecop from the Command Palette.

**Comment by ben-sussman-zocdoc** [comment]
No, that is "human experience time", measured from the time I click the `misc.txt` file (or launch it with Quicksilver, or launch it from the terminal) and the time that I see text displayed.

Timecop reports a significantly faster time of just under 1000ms. However it must be ignoring something because I experience >1000ms of time staring at a blank atom.io screen before the tab appears with my `txt` file in it.

**Comment by Gabriel-p** [comment]
Yeah I get a similar experience to that of @ben-sussman-zocdoc. I just tried: atom took 18s to fully load (really, it did, I used my cell's chronometer) while Timecop reported 9.8s.

After a couple of close/open cycles, I can get it down to 5.5s vs 1.4s (chronometer vs Timecop).

**Comment by Soleone** [comment]
on my rails project with around 40,000 files in total:

![image](https://cloud.githubusercontent.com/assets/2923/8485199/dd9cf03c-20c9-11e5-8370-ba9fdd38f619.png)

**Comment by wbyoung** [comment]
I'm seeing the same thing as @ben-sussman-zocdoc where human experience time is just a bit over 2x what Timecop reports (measured with stopwatch).

![captura de pantalla 2015-07-02 a las 1 55 03 pm](https://cloud.githubusercontent.com/assets/57162/8487607/1191313a-20c2-11e5-85c3-285c703c4716.png)

**Comment by thedaniel** [comment]
It's worth filing an issue on the timecop repo about this - if it's missing something that should be measured, it's less useful as a tool for us to diagnose problems.

**Comment by benogle** [comment]
Agreed on the timecop issue. Just speculating, but It's probably that something is happening before we can even start the clock. The reported time is at the beginning of the 'renderer' js code execution. See https://github.com/atom/atom/blob/master/static/index.js#L11 There is probably fruit to pick in the browser process as well.

**Comment by nemoDreamer** [comment]
:+1: All the devs I've talked to who didn't switch to Atom said it was because of the speed...
I adore Atom, but I have Sublime Text 3 set up as my default shell editor, so that I don't have to wait 3s+ to edit piped output, git commit messages, etc... I've got way more packages active in Sublime, but it still opens up like TextEdit (Notepad). Also it doesn't choke when editing 2000+ lines (but that's probably more the fault of plug-in developers...)

Sorry I'm not adding anything helpful to the conversation, and there _was_ a big improvement with 1.0, thanks!

**Comment by xpol** [comment]
Atom 1.0.9 on Mac mini late 2012 (16G RAM, HDD) still takes 3390ms for first time launch.

**Comment by mark-hahn** [comment]
How long does it take in safe mode?  (Start with "atom --safe")

**Comment by mnn** [comment]
With similar addons it's 27s vs. 3s. This is beyond a threshold of usability :(.

**Comment by danielmelogpi** [comment]
I'm consistently experiencing this problem, especially after a full reboot.
I can also have a very slow window reload if I delete compile-cache (understandable). I'm wondering if the use of such cache could be improved. I feel like after every reboot atom needs to do this again. If that is the case, we are missing the point of using a cache.

This was the startup time this morning:
![atom-superlento](https://cloud.githubusercontent.com/assets/3124696/10452774/8ae294e6-717d-11e5-82cb-be48851921a5.PNG)

**Comment by modest** [comment]
Let's say that there are 2 separate issues:
**Issue A:** Time to start Atom initially, issue #3673
**Issue B:** Time to create new editor window, **this issue**

Issue B is actually what stops me from using Atom very frequently. I use tabs sometimes, but I need multiple windows.  And new windows are insanely slow to create in Atom.

Why doesn't Atom just create an invisible "hot spare"?  Always have an extra Chromium window ready for action, with Atom loaded up, completely hidden to the user.  And when the user opens a document in a new window or creates a new editor window... it appears in milliseconds.

Yes, it's a little extra memory usage, but well worth it for the performance gain.  Atom is great at a lot of things, but it's not the right editor to choose if you are super tight on memory.

**Comment by rattrayalex** [comment]
FWIW, as a current non-user waiting solely for this issue to be resolved before becoming a user, I would happily pay the price of the extra ram in order to be able to use this software. 

+1 to @modest / @DavidLGoldberg / @blixt 's proposal

**Comment by iongion** [comment]
I have 32 GB of RAM, can't it be made to stay there and fire quickly ?
jEdit was doing such a thing and it shined(for startup-time compared to other java editors) in the early 2000s

**Comment by mark-hahn** [comment]
> can't it be made to stay there and fire quickly ?

Can't you just leave it open all the time?  That is the same thing.  I
never close mine.

**Comment by rattrayalex** [comment]
@mark-hahn I believe the issue is opening a new window, eg; for a separate project from the one you're already working on.

**Comment by amanse628** [comment]
Using samsung 850 evo 500GB SSD and it loads for 1461-2000 ms. Other programs on my SSD starts up fast. Is this normal?

**Comment by kolis** [comment]
I also have a fast workstation with lots of RAM and SSD, and Window load time is around 1500ms, using Atom 1.2 on Mac OS X.

**Comment by benogle** [comment]
The 1.3-beta that is out now has startup speed improvements see the [blog post](http://blog.atom.io/2015/11/12/atom-1-2.html) for more info. You can download the beta [here](https://atom.io/beta).

**Comment by modest** [comment]
Atom 1.3-beta here, still takes around 2000 ms to create a new window :(

**Comment by Gabriel-p** [comment]
Sadly I see very little improvement with 1.3-beta. From cold boot to fully open it just took over 7 sec. After an open/reopen cycle I can bring it down to 2200 ms.

Sublime 3 is almost instant in my system, for comparison.

**Comment by sliceofcode** [comment]
For me, it takes about ~2000ms (~1600 on beta) to create a window. It has improved, though. On previous versions of Atom, it took about 6-10 seconds to create a new window. (My computer is *not* fast.)

(1.3-beta on left, 1.2.4 on right)
![atom_compare](https://cloud.githubusercontent.com/assets/4206232/11429744/49b18f0a-9431-11e5-9d33-9cfe777520dc.png)

**Comment by agauniyal** [comment]
![2016-01-15-180750_1366x768_scrot](https://cloud.githubusercontent.com/assets/7630575/12353219/ebf989fe-bbb2-11e5-86f2-9a85d5eef204.png)

Says less than 2000ms but loads up completely in more than 5 seconds for me, even after disabling and uninstalling few essential packages.

**Comment by BladeMight** [comment]
8 sec+, why ?

**Comment by grinderX19** [comment]
Maybe it makes sense to implement something like LibreOffice's Systray Quickstarter as a workaround? It would be disabled by default.
I don't use such options for office apps as I don't work with documents much but it doesn't seem like a bad idea for Atom I use daily.

**Comment by alexandernst** [comment]
Doing what I wrote here https://github.com/atom/atom/issues/9334 would improve the situation.

**Comment by BladeMight** [comment]
@grinderX19, maybe, but what about how it does Visual Studio Quick Startup fucntion, without tray.

**Comment by BladeMight** [comment]
Or since it is using same framework that uses Chrome, why not to do like Chrome it does?
Chrome opens lighning fast, Atom(which uses same framework) opens tortoise like? This must me solved.
Or since it is using same framework that uses Chrome, why not to do like Chrome it does?
Chrome opens lighning fast, Atom(which uses same framework) opens tortoise like? This must be solved anyway.

**Comment by nathansobo** [comment]
I think there's something dicey going on in the browser process, maybe related to our outdated approach to restricting the application to a single instance, but it's just a hunch. Problem is our browser process code has evolved from a pile of Objective-C in the days when Atom lived in a WebKit window way before Electron even existed. It needs an overhaul, unit tests, etc. It's on my radar once I overhaul our outdated build system.

**Comment by alexandernst** [comment]
@nathansobo What do you think about my suggestion in #9334. Would that help? Is it possible at all?

**Comment by nathansobo** [comment]
@alexandernst I replied on that issue. Long story short, I don't think there's a silver bullet here. We just need to chip away. Straighten out the browser process and make sure it introduces no pauses and minimize the code we require. I have no better ideas at the moment.

**Comment by The-Penultimate-Defenestrator** [comment]
My atom is *really* slow to startup on a 2010 MacBook Pro. Seems like it takes ~6 seconds to actually show me anything, although a blank window spawns pretty quickly. Timecop reports ~4 seconds:
![](http://i.imgur.com/BYy2Kas.png)

**Comment by shellscape** [comment]
Mind bottling (yes, [bottling](https://www.youtube.com/watch?v=0DFBoLZC3Bw)) that this is still an open issue almost two years after it's been reported. Was googling for how to speed it up after it took nearly 10 seconds to open on my 2013 MBP tonight.

**Comment by mark-hahn** [comment]
>  this is still an open issue almost two years after it's been reported.

This isn't a bug to fix.  It is a performance improvement issue which will
get better and better but never go away.  Mine always opens in my laptop in
less than two seconds with about 40 packages but it will be better when it
opens in one second.  Two years ago Atom took six or more seconds for me to
open.  But even then it was worth it.

**Comment by agauniyal** [comment]
maybe this editor is designed to work on latest & resource heavy machines only, and we're just hoping that this will work fast enough on our 2-3yrs old systems. Timecop shows 2-3seconds startup but editor opens up in ~8 seconds on my machine.

**Comment by shellscape** [comment]
echoing the sentiments of @agauniyal. that's an excellent point about chrome's lightning quick startup by comparison. perhaps the atom team could reach out to the chrome developers group for some assistance - they might have the magic bullet.

**Comment by thomasjo** [comment]
> maybe this editor is designed to work on latest & resource heavy machines only, and we're just hoping that this will work fast enough on our 2-3yrs old systems.

I'm using a mid-2012 15" MBP (Retina) machine, hence an almost 4 year old machine, and have never had any major performance issues. Atom does a fair bit of disk IO during startup, so using a good SSD instead of an antiquated, spinning rust drive does help a ton though.

Comparing Atom to Chrome is like comparing apples and kittens; it doesn't really make sense because they are trying to achieve very different things. What Chrome does is trivial in comparison to Atom (since the latter is built directly on top of the former).

> perhaps the atom team could reach out to the chrome developers group for some assistance - they might have the magic bullet.

They have been involved several times, and let me assure you, there is no silver bullet. There are a multitude of many small improvements that can, and will, be made, that will gradually keep improving the performance. Performance has already improved a lot in the last year, and will likely continue to improve this year.

**Comment by prijindal** [comment]
I have something in my atom startup
It opens really fast in less than 2 seconds when i am opening a simple folder
But any folder which has npm modules takes considerable time to load.
What exactly Atom does when there are npm modules

**Comment by The-Penultimate-Defenestrator** [comment]
@thomasjo Yeah, but I'm using a mid-2010 MBP. I'm two years behind you, and I *do* have any major performance issues. I've got an SSD, though.

One thing I've been noticing lately is that atom slows down a *lot* when checking for updates. Even if I close the settings tab, it's almost unusably slow for about 2 minutes afterwards.

**Comment by thomasjo** [comment]
@The-Penultimate-Defenestrator I'm guessing what you're experiencing is related to https://github.com/atom/settings-view/issues/756.

**Comment by The-Penultimate-Defenestrator** [comment]
@thomasjo No, I see a huge slow-down even when searching for new packages, not just while updating.

**Comment by thomasjo** [comment]
> I have something in my atom startup
> It opens really fast in less than 2 seconds when i am opening a simple folder
> But any folder which has npm modules takes considerable time to load.
> What exactly Atom does when there are npm modules?

@prijindal It doesn't do anything special with npm modules, but Atom does run through the entire directory tree of your project, and if your project has a wide and/or deep `node_modules` directory tree with lots of files, scanning will spend a nontrivial amount of time and resources. If you e.g. add `node_modules` to your project's `.gitignore` file, your problem should go away (assuming you've not changed the default ignored files setting in Atom).

**Comment by Ehekatl** [comment]
Supper slow on windows 10...I have stand this since atom 1.0, a brand new atom is very fast at beginning, but suddenly without install any package it start become very slow, even in safe mode
![slow](http://7xicnm.com1.z0.glb.clouddn.com/slow.PNG)
![slow2](http://7xicnm.com1.z0.glb.clouddn.com/slow2.PNG)

**Comment by denji** [comment]
![timecop](https://cloud.githubusercontent.com/assets/3759759/14774887/90dce650-0ac3-11e6-8b6e-036e835ac8c1.png)

**Comment by Zireael07** [comment]
Is your atom storage folder large? Maybe something in there is slowing things down... a large project?

**Comment by lee-dohm** [comment]
@Zireael07 The `~/.atom/storage` folder isn't used in the same ways in v1.7.x and above.

**Comment by sryze** [comment]
Mine loads literally half a minute :) (Atom 1.7.3)

![screen shot 2016-05-18 at 15 07 27](https://cloud.githubusercontent.com/assets/4525736/15353381/f7984ba8-1d12-11e6-8c5f-ee9ac0277869.png)

It's a pretty old Mac though.

**Comment by muchweb** [comment]
If you have lots of packages installed, you may try storing your `.atom` directory in a RAM disk. This could make reopening a bit faster. I have my `/tmp` directory mounted as a `tmpfs`.
```bash
mv ~/.atom{,-storage}
ln -s ~/.atom /tmp/.atom

// On system start-up
cp -r ~/.atom-storage /tmp/.atom

// To preserve changes (e. g. updating modules)
rsync --itemize-changes --archive --delete /tmp/.atom/ ~/.atom-storage
```

**Comment by 4p** [comment]
@muchweb That work for me. Thanks

**Comment by sliceofcode** [comment]
Startup has definitely improved from Nov 2015, however this is a new laptop (1.9ghz i3)

![2016-06-08-16 23 33](https://cloud.githubusercontent.com/assets/4206232/15914385/5e4ab25e-2d95-11e6-99c7-fdee12347274.png)

**Comment by avm99963** [comment]
For me Atom lasted 20 seconds or more to load. But after following the steps indicated by thedaniel and thomasjo the time was reduced to 440ms, which is great hehe.

Basically what they mentioned (and I did) was removing the `~/.atom/storage/` folder and running `apm rebuild-module-cache`.

**Comment by denji** [comment]
> If you have lots of packages installed, you may try storing your .atom directory in a RAM disk.

NVMe SSDs 100k IOPS?

**Comment by BladeMight** [comment]
After updating from version 1.5.0(which i was using for long time) to latest 1.9.0beta1, startup become 8x faster, with same packages but startup faster ~1 sec. (AMD 2x2.5 GHz)

**Comment by dthorsen** [comment]
When I open Atom 1.11.12 from the Dock on Mac OS 10.11.6 on a brand new MacBook Pro, it takes 10's of seconds to load on Window load time.  
<img width="964" alt="screen shot 2016-11-01 at 11 02 42 am" src="https://cloud.githubusercontent.com/assets/10946253/19896586/dbf6f7c6-a022-11e6-9e69-93e327f2973b.png">

If I open it from terminal, it opens in 1 or 2 seconds.  What would cause this?
<img width="963" alt="screen shot 2016-11-01 at 11 02 16 am" src="https://cloud.githubusercontent.com/assets/10946253/19896592/e20eab2c-a022-11e6-89b7-00615dd6cbed.png">

**Comment by nathansobo** [comment]
@dthorsen It's caused by macOS's app nap feature. Basically the running instance of Atom goes to sleep and we can't communicate with it when a second instance is launched via the terminal. We have a fix in place on beta which should be out next week.

**Comment by torn4dom4n** [comment]
When will the faster?
Windows 10 Anniversary.
Ram 2Gb.
Intel Core i3.
Atom 1.12.3.

![atom-timecop](https://cloud.githubusercontent.com/assets/16760489/20343322/75c85396-ac20-11e6-821a-c4ac9ed061f0.png)

**Comment by nathansobo** [comment]
We're actively working on improving this but it will be a couple releases at least probably.

**Comment by torn4dom4n** [comment]
@nathansobo I see that GitKraken and VSCode use Electron but both activities seem faster and easier to install. I hope Atom will be better. I really like Atom.

**Comment by nathansobo** [comment]
Yeah, it has a lot more to do with the modular nature of Atom and the fact that we're super dynamic and flexible... lots of little paper cuts have built up over time. We'll have a tracking issue diagnosing the situation pretty soon.

**Comment by martinheidegger** [comment]
I am just putting this out there but for plugin development a lint tool [like I outlined here](https://github.com/nodejs/help/issues/304) could be used to check plugins and it would probably make the startup quite a bit faster: 

> It should check if my package is loaded quickly.

**Comment by 50Wliu** [comment]
> It should check if my package is loaded quickly.

Timecop already implements this.

**Comment by martinheidegger** [comment]
there is a surprising amount of packages called timecop. Which one are you referring to?

**Comment by Ben3eeE** [comment]
It is a core package. You can access it via the command `timecop:view` from the command palette.

**Comment by walles** [comment]
Related: https://github.com/atom/metrics/issues/70

**Comment by Xaeroxe23** [comment]
So pretty early on in this issue a distinction between Atom being an IDE and being a text editor was made.  For example when I launch Visual Studio I expect it to take a while to come up and in exchange for that I get a full IDE.  When I just need to review a log file though without launching a code editor programs like Notepad++ are relatively instantaneous and give me what I need quickly.

I use Atom to edit code written in the Rust programming language, and as a consequence of this I have Atom setup with a collection of packages that make it feel more like an IDE, and I can clearly see in my timecop that these packages are costing me startup time.  What I would like to be able to do but see no way to do in Atom as it currently exists is to have different "modes" with different lists of packages to load.  For my scenario I'd have two "modes" a default mode, which doesn't load any of the Rust "IDE" packages I use and a "Rust" mode which would load the Rust IDE packages.  Perhaps Atom could respond to a command line switch on launch something like "--mode=Rust" or perhaps after the default mode is launched you could have a "Mode" dropdown in the menu bar which lets you select a particular mode.  These modes would have to be user defined of course.

**Comment by Xaeroxe23** [comment]
Also, is Atom using this V8 functionality to its full extent? http://v8project.blogspot.com/2015/07/code-caching.html If not, caching the compilation results of packages and using them if the package hasn't been modified could help with startup time.

**Comment by prijindal** [comment]
@Xaeroxe23 that is a very cool feature. Something like that will also be helpful if you are using atom for two different programming languages or for two different frameworks.
Then you can simply save your configurations and plugins that need to load for that language and will reduce the startup time considerably, I suppose

**Comment by mildocjr** [comment]
A few years later and still wish Atom would launch faster.  On Windows 10 64-bit, with a quad core i7 @2.5Ghz 16 GB of RAM, from taskbar icon click to initial window launch it takes 8 seconds. From window launch to useable it takes another 4.5 seconds (12.5 seconds in total) from a near fresh install.  Would it be possible to rework the initial launch to load the window and document first, followed by the specific plug-ins needed for syntax highlighting and auto-completion?  The source tree could be loaded next, followed by any plug-ins that associate with other files in the source tree, and finally the remaining plug-ins that may not be used at all.  Atom has been my favorite code editor since I found out about it a few years ago.  The only reason I install Notepad++ is for previewing files quickly because of Atom's launch speed.

![image](https://cloud.githubusercontent.com/assets/8762467/21911636/3a1a6ffc-d8e7-11e6-8446-6f56e9218d0f.png)

**Comment by nathansobo** [comment]
Current work on launch speed is focused on snapshotting core code and bundled packages, initially to reduce time spent requiring JS but subsequently to remove other overhead. We need to modernize multiple packages in order to include them in the snapshot and you can track that work in https://github.com/atom/atom/issues/13254.

Short answer, we're working on it but there's no immediate silver bullet. Startup time is influenced by multiple factors.

**Comment by seanfuture** [comment]
Would it be possible to completely defer project loading and window state?  ( assuming shell loading is required, but even defer that too if possible ) .. Often times we're simply opening a single file and need to be able to access it as quickly as possible.  Right now Atom opens in about 2 seconds on a modern high-end MacBook and these 3 aspects as reported by Timecop seem to be the sizable portion.  It would be huge if async processing of these major items could happen after Atom displays, file is opened and visibly accessible.  Thanks for your insight @nathansobo.

![image](https://user-images.githubusercontent.com/5050950/27290317-d2d8f1f4-54da-11e7-8123-606aae46f920.png)

**Comment by nathansobo** [comment]
@seanfuture The devil is always in the details. These may be good ideas but would take time to investigate, so I can't really give you any specific insight. We'll circle back around to startup time eventually. In the meantime if you'd like to experiment with these ideas I'd be interested in your findings.

**Comment by gocarlos** [comment]
Is it not possible to put atom into the systemtray as telegram desktop or chrome does?

I work on linux, but like very much the apple behavior when closing a window, the app is still there but not visible in the overview... 

I think that by putting atom into the system tray after the first start, it could be lighting fast with lesser effort that taking other measures, I think it could also probably be done pretty easy as chrome does it already..

What do you think @nathansobo ?

**Comment by rafrafek** [comment]
Atom 1.28.1 startup is so slow it takes about 8-10 seconds on SSD with project like 30 lines of html and without any javascript (main.js on my screenshot is an empty file). I don't have any non-core package installed. 
For example VS Code starts < 1s on my machine. (SSD SanDisk Ultra II 240GB Sata3, Ryzen 1600 x6 core, 16 GB ram). 
I think something is really really wrong with Atom startup time. 
![atom_slow_startup_1-28-1](https://user-images.githubusercontent.com/23004737/42630041-be22ab6a-85d5-11e8-9f63-47d993530180.PNG)

**Comment by Jacalz** [comment]
To be quite honest, this is still really slow. Would like to see this improved dramatically, just look at vscode, it has way better startup time...

**Comment by benwhalley** [comment]
It might be worth keeping this issue open as startup time is still not that snappy (latest build, OS X).

**Comment by MartinX3** [comment]
Yes, it's way too much. (Latest build, Arch Linux)

**Comment by boredhero** [comment]
Yeah, I'd just like to bump this as well and add my +1 (Latest build, Arch Linux). It's ridiculous that with an SSD, Ryzen 3600, Fury X, and 16GB of ram it takes ~5-7 seconds to launch atom. I primarily use atom to edit small files when nano won't cut it, but

**Comment by ThatXliner** [comment]
Same. I use vim for extremely snappy edits, VS Code for debugging, and Atom for everything else. I hate vim. I really want to replace it with Atom. Is the init script execution already asynchronous?

**Comment by ventr1x** [comment]
5-7 seconds? On a maxed out Mac M1 it takes about 30 seconds. That's what we get for the decline of competency and the rise of cheap, fast packed web apps. Even in software dev people keep killing quality over easier and faster profits.

**Comment by Xaeroxe** [comment]
@ventr1x that's running emulated. Please reconsider your tone when making comments, you're not making friends or progress with that attitude.

**Comment by MartinX3** [comment]
@Xaeroxe In our opinion these multi platform electron website apps are a step backwards in terms of cpu, ram, disk storage and battery usage. Electron apps running their own chromium browser.
But hey, they are multi platform and html and javascript, etc.

**Comment by Xaeroxe** [comment]
I'm not here to talk about the performance of electron and whether it's good or bad. I'm here to remind people that

1. Open source maintainers owe you nothing.
2. Wandering into an issue tracker and calling everyone incompetent is never useful.

If you don't like Electron, why are you here? There are non-Electron editors out there, and Atom is always going to be an Electron product. Changing that would require starting from scratch, at which point you're no longer hacking on Atom. I don't know what you're trying to accomplish by bringing the Electron discussion here.

**Comment by MartinX3** [comment]
I dislike safe-spaces, they kill progress of any kind.
So I write critics about stuff I tried, but not anymore, because of its disadvantages.

**Comment by ventr1x** [comment]
It's not running "emulated".
Electron supports M1 for over a year "natively" now.
The same issue exists on my I9 MBP 2018.
Other Electron apps (including ones I developed) do not, even remotely, have such issues.

Issues without solution or disccusions - instead simple closures -, existing for 7+ years.. yep, I cannot call that a beacon of competence.

**Comment by icecream17** [comment]
It's more like there's not enough contributors. There's 7 pull requests right now and VSCode had 30+ commits today, so that's not a fair comparison. 7+ years is totally normal. In fact, I chose a random number between 1 and 100000, and this vscode issue took 1 year <sup>oof</sup> to resolve (probably worse than average but still) https://github.com/microsoft/vscode/issues/28965.

/rant (really unproductive comments here huh)

Anyways...
![image](https://user-images.githubusercontent.com/58114641/135353884-c80912ce-bdb2-4b23-9f29-dbf3340d1f35.png)

I also used the javascript profiler in the console, and here's the results when I did the reload window command:
![image](https://user-images.githubusercontent.com/58114641/135354450-cdc17a57-81e4-4016-89f4-b6aa755c704b.png)
